java/google/registry/whois/WhoisCommandFactory.java

```diff
@@ -0,0 +1,53 @@
+// Copyright 2016 The Nomulus Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package google.registry.whois;
+
+import com.google.common.net.InternetDomainName;
+import google.registry.config.RegistryConfig.ConfigModule;
+
+import javax.annotation.Nullable;
+import java.net.InetAddress;
+
+/**
+ * A no-op base class used to provide custom whois commands.
+ * <p>To add custom commands, extend this class, then configure it in
+ * {@link ConfigModule#provideWhoisCommandFactoryClass()}.
+ */
+public class WhoisCommandFactory {
+
+  public final WhoisCommand domainLookup(InternetDomainName domainName) {
+    return domainLookup(domainName, null);
+  }
+
+  public WhoisCommand domainLookup(InternetDomainName domainName, @Nullable InternetDomainName tld) {
```

Comment:
The tld arguments to these commands are a bit of a holdover from older days and aren't really relevant any more.  The only thing they accomplish is indicating to the domain and nameserverbyhost lookup commands that we already found a valid TLD so that they don't duplicate effort looking up the TLD again.

However, as written, those commands are still defensively validating the TLD if passed, even though that's duplicate effort, given that the only callsites pass in TLDs that must be valid.  And looking up the TLD from a given name is pretty cheap, and the marginal cost of doing the lookup over doing validation is very low.  So I think it'd be simpler overall to drop all the TLD parameters and just have the domain and nameserver commands always do their own TLD lookup.

That way we avoid needing two extra extension points in this API.

If you prefer, we can do the cleanup as a separate CL after this goes in, but just as a heads-up that this API would be changing if we do then make that cleanup.